iroh/src/magicsock/transports.rs

```diff
@@ -0,0 +1,432 @@
+            ip,
+            relay,
+        }
+    }
+
+    /// Send the given [`Transmit`] to the given [`Addr`].
+    ///
+    /// Sends on the first matching & ready transport.
+    pub(crate) fn poll_send(
+        &self,
+        destination: &Addr,
+        transmit: &Transmit<'_>,
+    ) -> Poll<io::Result<()>> {
+        trace!(?destination, "sending");
+
+        match destination {
+            #[cfg(wasm_browser)]
+            Addr::Ipv4(..) => {
+                return Poll::Ready(Err(io::Error::other("IPv4 is unsupported in browser")))
+            }
+            #[cfg(not(wasm_browser))]
+            Addr::Ipv4(addr, port) => {
+                let addr = SocketAddr::V4(SocketAddrV4::new(*addr, port.unwrap_or_default()));
+                for transport in &self.ip {
+                    if transport.is_valid_send_addr(&addr) {
+                        match transport.poll_send(addr, transmit) {
+                            Poll::Pending => {}
+                            Poll::Ready(res) => return Poll::Ready(res),
+                        }
+                    }
+                }
+            }
+            #[cfg(wasm_browser)]
+            Addr::Ipv6(..) => {
+                return Poll::Ready(Err(io::Error::other("IPv6 is unsupported in browser")))
+            }
+            #[cfg(not(wasm_browser))]
+            Addr::Ipv6(addr, port) => {
+                let addr = SocketAddr::V6(SocketAddrV6::new(*addr, port.unwrap_or_default(), 0, 0));
+                for transport in &self.ip {
+                    if transport.is_valid_send_addr(&addr) {
+                        match transport.poll_send(addr, transmit) {
+                            Poll::Pending => {}
+                            Poll::Ready(res) => return Poll::Ready(res),
+                        }
+                    }
+                }
+            }
+            Addr::RelayUrl(url, node_id) => {
+                for transport in &self.relay {
+                    if transport.is_valid_send_addr(url, node_id) {
+                        match transport.poll_send(url.clone(), *node_id, transmit) {
+                            Poll::Pending => {}
+                            Poll::Ready(res) => return Poll::Ready(res),
+                        }
+                    }
+                }
+            }
+        }
+        Poll::Pending
+    }
+
+    /// Tries to recv data, on all available transports.
+    pub(crate) fn poll_recv(
+        &self,
+        cx: &mut Context,
+        bufs: &mut [IoSliceMut<'_>],
+        metas: &mut [quinn_udp::RecvMeta],
+        source_addrs: &mut [Addr],
+    ) -> Poll<io::Result<usize>> {
+        debug_assert_eq!(bufs.len(), metas.len(), "non matching bufs & metas");
+
+        // TODO: randomization
+        macro_rules! poll_transport {
+            ($socket:expr) => {
+                match $socket.poll_recv(cx, bufs, metas, source_addrs)? {
+                    Poll::Pending | Poll::Ready(0) => {}
```

Comment:
Here I'm also a bit worried about `Ready(0)`.

Reply:
as above, this is what we have on `main`